### ISSUE TITLE
How to predict input shape of (depth*N, width, channel)?

### ISSUE THREAD

**Issue by gajop** [Dobiasd/frugally-deep#56]
In my use case, I have a tensor with `(depth*N, width, channel)` shape and I would like to use `predict` or `predict_multi` and get `(d_out*N, ...)` shape of output. Right now I need to do a lot of reshaping which makes for quite ugly code (shown below).
I was a bit surprised with how fdeep::tensor3_to_depth_slices works. 

Any suggestions how this code could be simplified?

```c++
auto vec_of_tensors = fdeep::tensor3_to_depth_slices(
	fdeep::internal::reshape_tensor3(t, std::vector<int>({ 1, 128, 64 }))
);
std::vector<fdeep::tensor3s> input(vec_of_tensors.size());
std::transform(vec_of_tensors.begin(), vec_of_tensors.end(), input.begin(),
	[] (const fdeep::tensor3& ts) -> fdeep::tensor3s { return fdeep::tensor3s({ ts }); });
```

**Comment by gajop** [comment]
I'm aware of the internal considerations, but thanks for the warning. I hope it further enforces the "there must be a better way" statement I was trying to make :) 

Q1. Technically, no. I want to convert it to `std::vector<fdeep::tensor3s>`, which is the input to `predict_multi`. The desire is to create a structure of fdeep::shape3(n * 1 * depth, height, width), where 1 is the number of inputs as I understand it.

Q2. Also no for the same reasons regarding input to `predict_multi`. While it would also be neat if output came as a `tensor3s` or `tensor3`, instead of vector of tensor3s, my particular use case makes it easy to deal with the output, since it's one-dimensional (or rather `n * (1, 1, 1)`).

I'm coming at this problem from Numpy thinking, where it's common to reshape and apply a function on a single axis. I believe keras also works in a similar way.

**Comment by Dobiasd** [comment]
OK, sorry, I temporarily forgot about the type signatures of `fdeep::predict` and `fdeep::predict_multi`.

So you don't want to use these "stacked" tensors for some kind of batch processing, but you have a model which takes multiples tensors of the same shape as input (for one single prediction) and also outputs multiple tensors of the same shape, is this correct?

In case I'm still not on the right track, could you please show me your model architecture? In case it is confidential perhaps just the input shapes and output shapes?

**Comment by gajop** [comment]
I *am* trying to achieve batch processing, at least as far as input and output shapes go. Maybe I miscommunicated something? I've been dealing with shape issues most of the day, and I am losing focus a bit :)

I have my own experimental architecture consisting of two models, as so:
```
Model 1:
[Input1] c: 1, w: 8192, h: 1 (not important)
[Input2] c: 1, w: 8192, h: 1 (not important)
[Input3] c: 64, w: 1, h: 1
[Input4] c: 64, w: 1, h: 1
[Output1] c: 8, w: 1, h: 1 (not important)

Model 2:
[Input5] c: 1, w: 128, h: 1
[Output2] c: 1, w: 1, h: 1
```
I reshape Input1 into (64, 128, 1) and feed it into Model 2, which gives me Input3 
I do the same with Input2, which results in Input4.
These two steps are what I'm trying to solve right now. Model2 normally takes (1, 128, 1) shape as input, but I would like to use it in batch processing and feed (64, 128, 1) as input and get (64, 1, 1) as output, which I could immediately feed to Model 1.

I then combine those first four inputs and feed them into Model1

**Comment by Dobiasd** [comment]
To me it feels like there could be a solution without so much manual wiring and reshaping by using [`keras.layers.Reshape`](https://keras.io/layers/core/#reshape).

However for the case that you want (or have to) wire up manually, let's try to solve it. Would having the following two functions help?

```cpp
// convert a vector with n tensors of shape (depth, height, width) into one tensor with shape (n * depth, height, width)
fdeep::tensor3 stack_tensor3s(const fdeep::tensor3s&);
```
and
```cpp
// convert a tensor with shape (n * depth, height, width) into a vector with n tensors of shape (depth, height, width).
fdeep::tensor3s unstack_tensor3s(std::size_t n, const fdeep::tensor3&);
```

**Comment by Dobiasd** [comment]
Mhh, perhaps there is another misunderstanding. In general frugally-deep does not support batch processing. `model::predict` only takes (and returns) multiple tensors because a Model created with the [functional API of Keras](https://keras.io/getting-started/functional-api-guide/) can have multiple input tensors (and output tensors).

`model::predict_multi` is just a `transform` of `predict` over multiple `tensor3s`.

**Comment by gajop** [comment]
I don't think `keras.layers.Reshape` is of exact use here, as the network is already properly accepting shapes like `(c: ?, w: 128, h: 1)` (at least in Python Keras it works). What I am trying to do is convert my `(c: 1, w: 128 * num_of_instances, h: 1)` into `(c: num_of_instances, w: 128, h: 1)`. 

Regarding the functions you proposed, as I understand it, `fdeep::tensor3s unstack_tensor3s` would work exactly the same as `fdeep::tensor3_to_depth_slices` for the special case where `depth = 1` (which is also my case)? While it would be useful, it still essentially produces a tensor3s, where I need a vector of tensor3s. 

The ugliest part of code in my previous examples is the below excerpt, where I am wrapping each of the tensor elements in a one-element vector (due to there being only one input):
```
std::transform(vec_of_tensors.begin(), vec_of_tensors.end(), input.begin(),
	[] (const fdeep::tensor3& ts) -> fdeep::tensor3s { return fdeep::tensor3s({ ts }); });
```
It is required because the functional API can take a variable number of inputs (which I use in Model1 quite well). 

> Mhh, perhaps there is another misunderstanding. In general frugally-deep does not support batch processing.

Maybe I'm misunderstanding something indeed, but isn't `model::predict_multi` essentially "batch processing"?

**Comment by Dobiasd** [comment]
Yes `model::predict_multi` does something that is a kind of batch processing, but not in the same way Keras does it, at least I think so. It is just running multiple `predict`s in parallel (multiple threads) without any fancy other optimizations that would reduce overall CPU time spent.

---

But regarding the grand theme of your problem: I'm a bit embarrassed, because I still don't understand why you are trying to do what you are trying to do. It feels like there must be a simpler solution.

However, perhaps I can help with the problem at hand without understanding the context nonetheless.

Let me try. :-)

So you have the following process:
```
shape notation: (channels, height, width)

1 tensor with shape (1, 1, 8192)
    -> reshape to 64 tensors of shape (1, 1, 128)
    -> model2.predict_multi
    -> get 64 tensors of shape (1, 1, 1)
    -> reshape to 1 tensor of shape (64, 1, 1)
    -> model1.predict
```

You have a working solution, but the code is ugly. The part you have shown is for the first reshape step in the chain above.

To make things at least a bit less ugly you could use `fplus::transform` instead of `std::transform`, i.e. instead of
```cpp
std::vector<fdeep::tensor3s> input(vec_of_tensors.size());
std::transform(vec_of_tensors.begin(), vec_of_tensors.end(), input.begin(),
    [] (const fdeep::tensor3& ts) -> fdeep::tensor3s { return fdeep::tensor3s({ ts }); });
```
this:
```cpp
const auto input = fplus::transform(
    fplus::singleton_seq<fdeep::tensor3>,
    vec_of_tensors);
```

`fplus::singleton_seq` is a function that converts `T` to `vector<T>` with one element.
You can find more of these functions here: http://www.editgym.com/fplus-api-search/

But the overall process would still not be very nice. The second reshape step would again be quite a mess.

I'll think about it and try to come up with a better solution.

**Comment by Dobiasd** [comment]
OK, I think I have something.

Let me first introduce some new notation:
- `(c, h, w)` -> shape for a `tensor3`
- `(v; c, h, w)` -> shape for a `std::vector<tensor3>`
- `(vv; v; c, h, w)` -> shape for a `std::vector<std::vector<tensor3>>`

If my understanding is correct, you need the following two reshape operations:
1) `(1, 1, 8192)` -> `(64; 1; 1, 1, 128)`
2) `(64; 1; 1, 1, 1)` -> `(1; 64, 1, 1)`

To remove the clutter from your code a function with the following signature would be good:

```cpp
tensor3s_vec reshape_tensor3_vectors(
    std::size_t vectors_size,
    std::size_t vector_size,
    std::size_t depth,
    std::size_t height,
    std::size_t width,
    const tensor3s_vec& tss);
```

It could then be used as follows:

```cpp
int main()
{
    const fdeep::tensor3 a(fdeep::shape3(1, 1, 8192), 0);

    // (1, 1, 8192) -> (64; 1; 1, 1, 128)
    const auto b = reshape_tensor3_vectors(64, 1, 1, 1, 128, {{a}});

    const auto c = model2.predict_multi(b);

    // (64; 1; 1, 1, 1) -> (1; 64, 1, 1)
    const auto d = reshape_tensor3_vectors(1, 64, 1, 1, 128, c).front();

    const auto e = model1.predict(d);
}
```

Please let me know if my understanding is correct and if the solution would be feasible for you. If so, I would implement `reshape_tensor3_vectors` and add it to the library.

**Comment by Dobiasd** [comment]
Just [added the discussed `reshape_tensor3_vectors` to `fdeep`](https://github.com/Dobiasd/frugally-deep/commit/1133c61c5730114bb649015fa40f22ec81c5d915).

To me it seems like this is a quite nice solution to your problem. Feel free to reopen the issue if it does not help in your use case.

**Comment by gajop** [comment]
Hello,
Sorry for the slow response. I was working on something else for a couple of days...

While I am grateful that you went as far as to implement this new functionality, I don't think it is ideal as far as the API design goes. It solves my very specific use case but it is not the obvious thing most users would pick. It seems to have a pretty complicated signature, and is quite likely to be avoided by users less familiar to this issue.

I think I had a problem with using the existing API because this library works different than what I would expect in Keras.
Instead of expecting a:
```
(N_batches, input0_dim1, input0_dim2, ... input0_dimn),  (N_batches, input1_dim1, input1_dim2, ... input1_dimn), .. (N_batches, inputK_dim1, inputK_dim2, ... inputK_dimn), 
```

It asks for a:
```
[ (input0_dim1, input0_dim2, ... input0_dimn),  (input1_dim1, input1_dim2, ... input1_dimn), .. (inputK_dim1, inputK_dim2, ... inputK_dimn), 

(input0_dim1, input0_dim2, ... input0_dimn),  (input1_dim1, input1_dim2, ... input1_dimn), .. (inputK_dim1, inputK_dim2, ... inputK_dimn),

... (N times)

(input0_dim1, input0_dim2, ... input0_dimn),  (input1_dim1, input1_dim2, ... input1_dimn), .. (inputK_dim1, inputK_dim2, ... inputK_dimn),
] 
```

For the moment, let's ignore if batches is the correct name for this. The main difference is that fdeep expects us to split the "vector of batched inputs" to a "batch of input vectors".

Generally life would be easier if predict_multi would just expect the input in the first form, and did the necessary reshaping/splitting within the implementation

**Comment by Dobiasd** [comment]
OK, so to represent `(N, C, H, W)` we would need a `tensor4` class.

Then we would have
```cpp
float tensor3::get(std::size_t z, std::size_t y, std::size_t x);
// tensor3::set accordingly
```
and
```cpp
float tensor4::get(std::size_t n, std::size_t z, std::size_t y, std::size_t x);
```

Currently we use `vector<tensor3>`, so instead of accessing it like
```cpp
tensor4 t;
t.get(n, z, y, x);
```

```cpp
tensor3s ts;
t[n].get(z, y, x);
```

which for me does not look that different. The order of dimensions is still the same.

Actually, one could argue the following
```
tensor1 === vector<float>
tensor2 === vector<vector<float>>
tensor3 === vector<vector<float>>>
tensor4 === vector<vector<vector<float>>>>
```

So currently I don't really see such a big difference between
```
(N, C, H, W)
```
and
```
[(C, H, W), (C, H, W), ... (C, H, W)]
```

---

I think it only gets really relevant when we enter the 5th dimension, i.e. when we consider batch prediction together with the possibility of models taking multiple tensors as input for a single prediction. In that case it would be
`(N_batches, N_tensors, C, H, W)` vs. `(N_tensors, N_batches, C, H, W)`
or in case of models taking/returning only one tensor
`(N_batches, 0, C, H, W)` vs. `(0, N_batches, C, H, W)`

---

But perhaps I again did not understand the problem fully, which seem to be a trend in this GitHub issue. :wink:

Would you mind proposing a type signature / function signature for the API what you think would be intuitive to use?
